### DISCUSSION TITLE
How to replicate AssertionError

### DISCUSSION THREAD

**Discussion by rad-pat** [sqlalchemy/sqlalchemy#6257] [Usage Questions]
I am receiving an AssertionError in 1.4.7, code works fine in 1.3.24. My attempts to build a minimal repl don't seem to yield the same error so I was wondering what I could gather to help debug this?

Traceback:
```
Traceback (most recent call last):
  File "/home/src/sql_expression.py", line 864, in build_query
    source_query.append_column(sqlalchemy.literal(1).label('flagged'))
  File "<string>", line 2, in append_column
  File "/home/plaid/py39/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 390, in warned
    return fn(*args, **kwargs)
  File "/home/plaid/py39/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 3985, in append_column
    self.add_columns.non_generative(self, column)
  File "/home/plaid/py39/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 5261, in add_columns
    self._assert_no_memoizations()
  File "/home/plaid/py39/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 1143, in _assert_no_memoizations
    assert elem not in self.__dict__
AssertionError
```

**Comment by zzzeek** [comment]
can you pdb in there, or do a print statement and show me the value of "elem" and `self.__dict__`?  that will allow me to see what's not being handled.

**Reply by rad-pat** [reply]
Yes, here are the requested values, looks like `_implicit_subquery` is causing the AssertionError. I'm pretty sure there should be no subquery involved here, it's just a straight-up select, no where clause either.

In[2]: elem
Out[2]: '_implicit_subquery'
In[3]: self.__dict__
Out[3]: 
{'_auto_correlate': True,
 '_raw_columns': [<sqlalchemy.sql.elements.Label object at 0x7f0152676430>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676520>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676610>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676700>,
  <sqlalchemy.sql.elements.Label object at 0x7f01526767f0>,
  <sqlalchemy.sql.elements.Label object at 0x7f01526768e0>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676a00>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676af0>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676be0>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676d30>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676e80>,
  <sqlalchemy.sql.elements.Label object at 0x7f0152676fd0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b160>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b250>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b340>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b430>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b520>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b610>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b700>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b7f0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b8e0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267b9d0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267bac0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267bbb0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267bca0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267bd90>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267be80>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267bf70>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e0a0>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e190>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e280>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e370>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e460>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e550>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e640>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e730>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e820>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267e910>,
  <sqlalchemy.sql.elements.Label object at 0x7f015267ea00>],
 '_label_style': symbol('LABEL_STYLE_DISAMBIGUATE_ONLY'),
 '_bind': None,
 '_where_criteria': (<sqlalchemy.sql.elements.BooleanClauseList object at 0x7f0152806580>,),
 '_implicit_subquery': <sqlalchemy.sql.selectable.Subquery at 0x7f015260beb0; anon_1>,
 '_memoized_keys': frozenset({'_implicit_subquery'})}

**Reply by zzzeek** [reply]
ah OK, I was wondering how you were getting that then I see you're using an API I thought nobody ever used anymore.....OK so there is a bug here in SQLAlchemy, which I will fix, however, on your end you are using at least two APIs that are deprecated:

1. append_column() - the select() construct works very poorly wrt in-place mutation, and this API will be removed in the next major SQLAlchemy release (either 1.5 or 2.0).     Generate a new select() using add_column()
2. "implicit subquery" is only present if you use the .c attribute or the select() method on your existing select() statement.  These two elements are also deprecated.  To access the columns that would be selected by your select(), use the .selected_columns attribute.


your program should be generating verbose warnings indicating both item 1 and 2 above,and these are independent of the new "2.0 warnings" mode - running your program with warnings enabled or set to error will reveal where these APIs are used  (i.e. -Walways or -Werror).

**Reply by zzzeek** [reply]
importantly, while I can probably fix this assertion, doing so is likely going to still produce a non-working result on your end because if you are appending columns, that "implicit subquery" that is being accessed wont be correspondingly updated

**Reply by rad-pat** [reply]
OK, many thanks for the detailed reply. I'll take a look at the code that builds this query and see what I can fix there. I certainly know where it is doing the append_column so I can fix that immediately. Interestingly, I thought I was running with warnings enabled, I certainly see other warnings such as for asyncio - I'll look into that too!!

**Reply by zzzeek** [reply]
if it's not generating the warnings then we have to solve that too , do you have any warnings.filterwarnings in your program ?

**Reply by rad-pat** [reply]
Well, there we go... I thought I had warnings enabled but I did not.  🤦 
I think those warnings will keep me busy a while!!!

**Reply by zzzeek** [reply]
https://github.com/sqlalchemy/sqlalchemy/issues/6261
